File: api/src/main/java/org/pmiops/workbench/db/model/DataDictionaryEntry.java
package org.pmiops.workbench.db.model;

import java.sql.Timestamp;
import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.ManyToOne;
import javax.persistence.Table;

@Entity
@Table(name = "data_dictionary_entry")
public class DataDictionaryEntry {

  // Meta data fields
  private long dataDictionaryEntryId;
  private CdrVersion cdrVersion;
  private Timestamp definedTime;

  // Fields copied from the Data Dictionary export
  private String relevantOmopTable;
  private String fieldName;
  private String omopCdmStandardOrCustomField;
  private String description;
  private String fieldType;
  private String dataProvenance;
  private String sourcePpiModule;
  private Boolean transformedByRegisteredTierPrivacyMethods;

  public DataDictionaryEntry() {}

  @Id
  @GeneratedValue(strategy = GenerationType.IDENTITY)
  @Column(name = "data_dictionary_entry_id")
  public long getDataDictionaryEntryId() {
    return dataDictionaryEntryId;
  }

  public void setDataDictionaryEntryId(long dataDictionaryEntryId) {
    this.dataDictionaryEntryId = dataDictionaryEntryId;
  }

  @ManyToOne
  @JoinColumn(name = "cdr_version_id")
  public CdrVersion getCdrVersion() {
    return cdrVersion;
  }

  public void setCdrVersion(CdrVersion cdrVersion) {
    this.cdrVersion = cdrVersion;
  }

  @Column(name = "defined_time")
  public Timestamp getDefinedTime() {
    return definedTime;
  }

  public void setDefinedTime(Timestamp definedTime) {
    this.definedTime = definedTime;
  }

  @Column(name = "relevant_omop_table")
  public String getRelevantOmopTable() {
    return relevantOmopTable;
  }

  public void setRelevantOmopTable(String relevantOmopTable) {
    this.relevantOmopTable = relevantOmopTable;
  }

  @Column(name = "field_name")
  public String getFieldName() {
    return fieldName;
  }

  public void setFieldName(String fieldName) {
    this.fieldName = fieldName;
  }

  @Column(name = "omop_cdm_standard_or_custom_field")
  public String getOmopCdmStandardOrCustomField() {
    return omopCdmStandardOrCustomField;
  }

  public void setOmopCdmStandardOrCustomField(String omopCdmStandardOrCustomField) {
    this.omopCdmStandardOrCustomField = omopCdmStandardOrCustomField;
  }

  @Column(name = "description")
  public String getDescription() {
    return description;
  }

  public void setDescription(String description) {
    this.description = description;
  }

  @Column(name = "field_type")
  public String getFieldType() {
    return fieldType;
  }

  public void setFieldType(String fieldType) {
    this.fieldType = fieldType;
  }

  @Column(name = "data_provenance")
  public String getDataProvenance() {
    return dataProvenance;
  }

  public void setDataProvenance(String dataProvenance) {
    this.dataProvenance = dataProvenance;
  }

  @Column(name = "source_ppi_module")
  public String getSourcePpiModule() {
    return sourcePpiModule;
  }

  public void setSourcePpiModule(String sourcePpiModule) {
    this.sourcePpiModule = sourcePpiModule;
  }

  @Column(name = "transformed_by_registered_tier_privacy_methods")
  public Boolean getTransformedByRegisteredTierPrivacyMethods() {
    return transformedByRegisteredTierPrivacyMethods;
  }

  public void setTransformedByRegisteredTierPrivacyMethods(
      Boolean transformedByRegisteredTierPrivacyMethods) {
    this.transformedByRegisteredTierPrivacyMethods = transformedByRegisteredTierPrivacyMethods;
  }
}

File: api/src/main/java/org/pmiops/workbench/db/model/DataDictionaryEntry.kt
package org.pmiops.workbench.db.model

import java.sql.Timestamp
import java.util.Objects
import javax.persistence.Column
import javax.persistence.Entity
import javax.persistence.GeneratedValue
import javax.persistence.GenerationType
import javax.persistence.Id
import javax.persistence.JoinColumn
import javax.persistence.ManyToOne
import javax.persistence.Table

@Entity
@Table(name = "data_dictionary_entry")
class DataDictionaryEntry {

    // Metadata fields
    @get:Id
    @get:GeneratedValue(strategy = GenerationType.IDENTITY)
    @get:Column(name = "data_dictionary_entry_id")
    var dataDictionaryEntryId: Long = 0
    @get:ManyToOne
    @get:JoinColumn(name = "cdr_version_id")
    var cdrVersion: CdrVersion? = null
    @get:Column(name = "defined_time")
    var definedTime: Timestamp? = null

    // Fields copied from the Data Dictionary export
    // See available_field_item in the following
    // https://github.com/all-of-us/cdrdatadictionary/blob/development/cdr_data_dictionary/schema.yaml
    @get:Column(name = "relevant_omop_table")
    var relevantOmopTable: String? = null
    @get:Column(name = "field_name")
    var fieldName: String? = null
    @get:Column(name = "omop_cdm_standard_or_custom_field")
    var omopCdmStandardOrCustomField: String? = null
    @get:Column(name = "description")
    var description: String? = null
    @get:Column(name = "field_type")
    var fieldType: String? = null
    @get:Column(name = "data_provenance")
    var dataProvenance: String? = null
    @get:Column(name = "source_ppi_module")
    var sourcePpiModule: String? = null
    @get:Column(name = "transformed_by_registered_tier_privacy_methods")
    var transformedByRegisteredTierPrivacyMethods: Boolean? = null

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false
        val that = o as DataDictionaryEntry?
        return (dataDictionaryEntryId == that!!.dataDictionaryEntryId
                && cdrVersion == that.cdrVersion
                && definedTime == that.definedTime
                && relevantOmopTable == that.relevantOmopTable
                && fieldName == that.fieldName
                && omopCdmStandardOrCustomField == that.omopCdmStandardOrCustomField
                && description == that.description
                && fieldType == that.fieldType
                && dataProvenance == that.dataProvenance
                && sourcePpiModule == that.sourcePpiModule
                && transformedByRegisteredTierPrivacyMethods == that.transformedByRegisteredTierPrivacyMethods)
    }

    override fun hashCode(): Int {
        return Objects.hash(
                dataDictionaryEntryId,
                cdrVersion,
                definedTime,
                relevantOmopTable,
                fieldName,
                omopCdmStandardOrCustomField,
                description,
                fieldType,
                dataProvenance,
                sourcePpiModule,
                transformedByRegisteredTierPrivacyMethods)
    }
}
